apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/asyncprofiler/AsyncProfilerTaskChannelService.java

```diff
@@ -0,0 +1,115 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.apache.skywalking.apm.agent.core.asyncprofiler;
+
+import io.grpc.Channel;
+import org.apache.skywalking.apm.agent.core.boot.BootService;
+import org.apache.skywalking.apm.agent.core.boot.DefaultImplementor;
+import org.apache.skywalking.apm.agent.core.boot.DefaultNamedThreadFactory;
+import org.apache.skywalking.apm.agent.core.boot.ServiceManager;
+import org.apache.skywalking.apm.agent.core.commands.CommandService;
+import org.apache.skywalking.apm.agent.core.conf.Config;
+import org.apache.skywalking.apm.agent.core.logging.api.ILog;
+import org.apache.skywalking.apm.agent.core.logging.api.LogManager;
+import org.apache.skywalking.apm.agent.core.remote.GRPCChannelListener;
+import org.apache.skywalking.apm.agent.core.remote.GRPCChannelManager;
+import org.apache.skywalking.apm.agent.core.remote.GRPCChannelStatus;
+import org.apache.skywalking.apm.network.common.v3.Commands;
+import org.apache.skywalking.apm.network.language.asyncprofiler.v10.AsyncProfilerTaskCommandQuery;
+import org.apache.skywalking.apm.network.language.asyncprofiler.v10.AsyncProfilerTaskGrpc;
+import org.apache.skywalking.apm.util.RunnableWithExceptionProtection;
+
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import static org.apache.skywalking.apm.agent.core.conf.Config.Collector.GRPC_UPSTREAM_TIMEOUT;
+
+@DefaultImplementor
+public class AsyncProfilerTaskChannelService implements BootService, Runnable, GRPCChannelListener {
+    private static final ILog LOGGER = LogManager.getLogger(AsyncProfilerTaskChannelService.class);
+
+    // channel status
+    private volatile GRPCChannelStatus status = GRPCChannelStatus.DISCONNECT;
+    private volatile AsyncProfilerTaskGrpc.AsyncProfilerTaskBlockingStub asyncProfilerTaskBlockingStub;
+
+    // query task list schedule
+    private volatile ScheduledFuture<?> getTaskListFuture;
+
+    @Override
+    public void run() {
+        if (status == GRPCChannelStatus.CONNECTED) {
+            // test start command and 10s after put stop command
+            long lastCommandCreateTime = ServiceManager.INSTANCE
+                    .findService(AsyncProfilerTaskExecutionService.class).getLastCommandCreateTime();
+
+            AsyncProfilerTaskCommandQuery query = AsyncProfilerTaskCommandQuery.newBuilder()
+                    .setServiceInstance(Config.Agent.INSTANCE_NAME)
+                    .setService(Config.Agent.SERVICE_NAME)
+                    .setLastCommandTime(lastCommandCreateTime)
+                    .build();
+            Commands commands = asyncProfilerTaskBlockingStub.withDeadlineAfter(GRPC_UPSTREAM_TIMEOUT, TimeUnit.SECONDS)
+                    .getAsyncProfilerTaskCommands(query);
+            ServiceManager.INSTANCE.findService(CommandService.class).receiveCommand(commands);
```

Comment:
You should check `service not implemented` error of gRPC, you could try that by connecting v < 10.1 OAP.

Reply:
Once it is not implemented, the agent should shutdown this part automatically.